k6 doesn't properly anonymize archives

When creating archives (i.e. when running the `k6 cloud` and `k6 arrchive` commands), k6 tries to anonymize the usernames in any paths that are `open()`-ed or imported: https://github.com/loadimpact/k6/blob/5788beef154b31f95ea8c6da5e86521b6fa07f6a/lib/archive.go#L38-L50

The problem comes from the fact that any file paths in the archives will be anonymized, but the paths in the actual `open("/home/username/file.json")` and `import test from "/home/my_other_username/superlib.js` won't... 

Example script:
```js
import mylib from "/home/user/mylib.js";

var someData = open("/home/user/data.json");

var username = "user" // Can't anonymize that without some strong AI or a magical JS parser :/ ...
var someMoreData = open("/home/" + username + "/data.json");

var theJoyOfJS = eval("open(['/home', 'user', 'data.json'].join('/'))");

export default function () {
    console.log(JSON.stringify(mylib));
    console.log(someData);
    console.log(someMoreData);
    console.log(theJoyOfJS);
}
```

As you might have deduced from the above example, I don't think this is fixable and we should just give up on the whole username anonymization idea...

The produced `archive.tar` file from running `k6 archive` on the above script results in a file that has a `data` file (i.e. the script entrypoint) exactly the same as the script above (no anonymization at all) and a tar file that looks like this (via `tar --list --file archive.tar`):
```  
metadata.json
data
scripts
scripts/_
scripts/_/home
scripts/_/home/nobody
scripts/_/home/nobody/mylib.js
files
files/_
files/_/home
files/_/home/nobody
files/_/home/nobody/data.json
```

This doesn't work, and what's more, `k6 run archive.tar` fails a panic, which is another bug in and of itself:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: Panic at 29: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: Panic at 29: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x6d97f5]

goroutine 1 [running]:
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunProgram.func1(0xc4231a2af8)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:880 +0x99
panic(0xbeb880, 0xc42307db20)
	/usr/lib/go/src/runtime/panic.go:502 +0x229
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try.func1(0xc4202d8d00, 0x0, 0xc4231a29f0, 0x0, 0x0, 0x0, 0xc4231a2a48)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:364 +0x45c
panic(0xc03940, 0x14cd250)
	/usr/lib/go/src/runtime/panic.go:502 +0x229
github.com/loadimpact/k6/vendor/github.com/spf13/afero.ReadFile(0x0, 0x0, 0xc420228714, 0x13, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/spf13/afero/ioutil.go:65 +0x45
github.com/loadimpact/k6/loader.Load(0x0, 0x0, 0xc4200364e0, 0x25, 0xc420228714, 0x13, 0x0, 0xc4231a21b8, 0x411cc9)
	/go/src/github.com/loadimpact/k6/loader/loader.go:100 +0x3ff
github.com/loadimpact/k6/js.(*InitContext).requireFile(0xc420c22780, 0xc420228714, 0x13, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/js/initcontext.go:135 +0x648
github.com/loadimpact/k6/js.(*InitContext).Require(0xc420c22780, 0xc420228714, 0x13, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/js/initcontext.go:97 +0xe7
reflect.callMethod(0xc42112e3c0, 0xc4231a2340)
	/usr/lib/go/src/reflect/value.go:663 +0x181
reflect.methodValueCall(0xc420228714, 0x13, 0x0, 0x0, 0xc4231a2620, 0x4b24e9, 0xc42112e330, 0xc42112e3c0, 0xc4208d1ac0, 0x1000000020, ...)
	/usr/lib/go/src/reflect/asm_amd64.s:29 +0x33
reflect.Value.call(0xbd5da0, 0xc42112e3c0, 0x13, 0xd13721, 0x4, 0xc4208d1aa0, 0x1, 0x1, 0x0, 0x0, ...)
	/usr/lib/go/src/reflect/value.go:447 +0x969
reflect.Value.Call(0xbd5da0, 0xc42112e3c0, 0x13, 0xc4208d1aa0, 0x1, 0x1, 0xc42307da80, 0x98, 0x0)
	/usr/lib/go/src/reflect/value.go:308 +0xa4
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).wrapReflectFunc.func1(0xe95b00, 0x1501960, 0xc4222c2220, 0x1, 0x1e, 0x7, 0x0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1153 +0x8c4
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm)._nativeCall(0xc4202d8d00, 0xc4208cd550, 0x1)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1826 +0x2ae
github.com/loadimpact/k6/vendor/github.com/dop251/goja.call.exec(0x1, 0xc4202d8d00)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1810 +0x51d
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).run(0xc4202d8d00)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:288 +0x51
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).(github.com/loadimpact/k6/vendor/github.com/dop251/goja.run)-fm()
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:375 +0x2a
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try(0xc4202d8d00, 0xc4231a2a50, 0x0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:370 +0x10c
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).runTry(0xc4202d8d00, 0xe1eee0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:375 +0x46
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunProgram(0xc42136e480, 0xc4215cf020, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:891 +0x1c2
github.com/loadimpact/k6/js.(*Bundle).instantiate(0xc42175f600, 0xc42136e480, 0xc420c22780, 0xc422b1e6ab, 0x1)
	/go/src/github.com/loadimpact/k6/js/bundle.go:262 +0x472
github.com/loadimpact/k6/js.(*Bundle).Instantiate(0xc42175f600, 0xbbb600, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/js/bundle.go:210 +0x15b
github.com/loadimpact/k6/js.(*Runner).newVU(0xc420ce1040, 0xc421ee8300, 0xc734e0, 0x462301, 0xc421faad40)
	/go/src/github.com/loadimpact/k6/js/runner.go:115 +0x46
github.com/loadimpact/k6/js.(*Runner).NewVU(0xc420ce1040, 0xc421ee8300, 0xc420b2db50, 0x1, 0xc420ac89b0, 0xc420c22230)
	/go/src/github.com/loadimpact/k6/js/runner.go:106 +0x35
github.com/loadimpact/k6/core/local.(*Executor).SetVUsMax(0xc420b2db00, 0x1, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/core/local/local.go:528 +0x34e
github.com/loadimpact/k6/core.NewEngine(0xe97420, 0xc420b2db00, 0x0, 0x1, 0x0, 0x1, 0x0, 0x0, 0x0, 0x1, ...)
	/go/src/github.com/loadimpact/k6/core/engine.go:86 +0x174
github.com/loadimpact/k6/cmd.glob..func11(0x14d8520, 0xc42056c4d0, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/loadimpact/k6/cmd/run.go:182 +0xa7b
github.com/loadimpact/k6/vendor/github.com/spf13/cobra.(*Command).execute(0x14d8520, 0xc42056c4a0, 0x1, 0x1, 0x14d8520, 0xc42056c4a0)
	/go/src/github.com/loadimpact/k6/vendor/github.com/spf13/cobra/command.go:762 +0x468
github.com/loadimpact/k6/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x14d82c0, 0x52, 0xc420576180, 0xc4205ce000)
	/go/src/github.com/loadimpact/k6/vendor/github.com/spf13/cobra/command.go:852 +0x30a
github.com/loadimpact/k6/vendor/github.com/spf13/cobra.(*Command).Execute(0x14d82c0, 0xc420287f78, 0x405d8c)
	/go/src/github.com/loadimpact/k6/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/loadimpact/k6/cmd.Execute()
	/go/src/github.com/loadimpact/k6/cmd/root.go:87 +0x31
main.main()
	/go/src/github.com/loadimpact/k6/main.go:28 +0x20
```

So... I think we should:
- disable that usename anonymization since apparently it has never worked and I don't see how we can make it work properly (t.e. by modifying the script as well as the file paths)
- actually use the `NormalizeAndAnonymizePath` (after removing `homeDirRE`) when we open/import files (when we `run` archives), since it's very likely that the other two regular expressions may also cause issues with the changed file paths...
- fix the panic - k6 shlouldn't panic, even if an archive file is wrong somehow
- test all of those things...

--- a/js/bundle.go
+++ b/js/bundle.go
@@ -23,6 +23,7 @@ package js
 import (
 	"context"
 	"encoding/json"
+	"os"
 
 	"github.com/dop251/goja"
 	"github.com/loadimpact/k6/js/common"
@@ -140,20 +141,8 @@ func NewBundleFromArchive(arc *lib.Archive, rtOpts lib.RuntimeOptions) (*Bundle,
 		return nil, errors.Errorf("expected bundle type 'js', got '%s'", arc.Type)
 	}
 
-	pgm, _, err := compiler.Compile(string(arc.Data), arc.Filename, "", "", true)
-	if err != nil {
-		return nil, err
-	}
-
-	initctx := NewInitContext(goja.New(), compiler, new(context.Context), nil, arc.Pwd)
-	for filename, data := range arc.Scripts {
-		src := string(data)
-		pgm, err := initctx.compileImport(src, filename)
-		if err != nil {
-			return nil, err
-		}
-		initctx.programs[filename] = programWithSource{pgm, src}
-	}
+	initctx := NewInitContext(goja.New(), compiler, new(context.Context), arc.FS, arc.Pwd)
+	pgm, err := initctx.compileImport(string(arc.Data), arc.Filename)
 	initctx.files = arc.Files
 
 	env := arc.Env
@@ -175,9 +164,10 @@ func NewBundleFromArchive(arc *lib.Archive, rtOpts lib.RuntimeOptions) (*Bundle,
 	}, nil
 }
 
-func (b *Bundle) MakeArchive() *lib.Archive {
+func (b *Bundle) makeArchive() *lib.Archive {
 	arc := &lib.Archive{
 		Type:     "js",
+		FS:       afero.NewMemMapFs(),
 		Options:  b.Options,
 		Filename: b.Filename,
 		Data:     []byte(b.Source),
@@ -192,6 +182,10 @@ func (b *Bundle) MakeArchive() *lib.Archive {
 	arc.Scripts = make(map[string][]byte, len(b.BaseInitContext.programs))
 	for name, pgm := range b.BaseInitContext.programs {
 		arc.Scripts[name] = []byte(pgm.src)
+		err := afero.WriteFile(arc.FS, name, []byte(pgm.src), os.ModePerm)
+		if err != nil {
+			return nil
+		}
 	}
 	arc.Files = b.BaseInitContext.files
 

--- a/js/bundle.go
+++ b/js/bundle.go
@@ -55,7 +55,7 @@ type BundleInstance struct {
 	Default goja.Callable
 }
 
-// Creates a new bundle from a source file and a filesystem.
+// NewBundle creates a new bundle from a source file and a filesystem.
 func NewBundle(src *lib.SourceData, fs afero.Fs, rtOpts lib.RuntimeOptions) (*Bundle, error) {
 	compiler, err := compiler.New()
 	if err != nil {
@@ -131,6 +131,7 @@ func NewBundle(src *lib.SourceData, fs afero.Fs, rtOpts lib.RuntimeOptions) (*Bu
 	return &bundle, nil
 }
 
+// NewBundleFromArchive creates a new bundle from an lib.Archive.
 func NewBundleFromArchive(arc *lib.Archive, rtOpts lib.RuntimeOptions) (*Bundle, error) {
 	compiler, err := compiler.New()
 	if err != nil {
@@ -143,6 +144,9 @@ func NewBundleFromArchive(arc *lib.Archive, rtOpts lib.RuntimeOptions) (*Bundle,
 
 	initctx := NewInitContext(goja.New(), compiler, new(context.Context), arc.FS, arc.Pwd)
 	pgm, err := initctx.compileImport(string(arc.Data), arc.Filename)
+	if err != nil {
+		return nil, err
+	}
 	initctx.files = arc.Files
 
 	env := arc.Env

--- a/js/bundle_test.go
+++ b/js/bundle_test.go
@@ -376,7 +376,7 @@ func TestNewBundleFromArchive(t *testing.T) {
 	}
 	assert.Equal(t, "hi!", v.Export())
 
-	arc := b.MakeArchive()
+	arc := b.makeArchive()
 	assert.Equal(t, "js", arc.Type)
 	assert.Equal(t, lib.Options{VUs: null.IntFrom(12345)}, arc.Options)
 	assert.Equal(t, "/path/to/script.js", arc.Filename)
@@ -456,7 +456,7 @@ func TestBundleEnv(t *testing.T) {
 		return
 	}
 
-	b2, err := NewBundleFromArchive(b1.MakeArchive(), lib.RuntimeOptions{})
+	b2, err := NewBundleFromArchive(b1.makeArchive(), lib.RuntimeOptions{})
 	if !assert.NoError(t, err) {
 		return
 	}

--- a/js/initcontext.go
+++ b/js/initcontext.go
@@ -74,8 +74,9 @@ func newBoundInitContext(base *InitContext, ctxPtr *context.Context, rt *goja.Ru
 		runtime: rt,
 		ctxPtr:  ctxPtr,
 
-		fs:  nil,
-		pwd: base.pwd,
+		fs:       base.fs,
+		pwd:      base.pwd,
+		compiler: base.compiler,
 
 		programs: base.programs,
 		files:    base.files,

--- a/js/runner.go
+++ b/js/runner.go
@@ -99,7 +99,7 @@ func NewFromBundle(b *Bundle) (*Runner, error) {
 }
 
 func (r *Runner) MakeArchive() *lib.Archive {
-	return r.Bundle.MakeArchive()
+	return r.Bundle.makeArchive()
 }
 
 func (r *Runner) NewVU(samplesOut chan<- stats.SampleContainer) (lib.VU, error) {

--- a/js/runner_test.go
+++ b/js/runner_test.go
@@ -1400,11 +1400,11 @@ func TestArchiveNotPanicing(t *testing.T) {
 	defer tb.Cleanup()
 
 	fs := afero.NewMemMapFs()
-	require.NoError(t, afero.WriteFile(fs, "/non/existant", []byte(`42`), os.ModePerm))
+	require.NoError(t, afero.WriteFile(fs, "/non/existent", []byte(`42`), os.ModePerm))
 	r1, err := New(&lib.SourceData{
 		Filename: "/script.js",
 		Data: []byte(tb.Replacer.Replace(`
-			let fput = open("/non/existant");
+			let fput = open("/non/existent");
 			export default function(data) {
 			}
 		`)),

--- a/js/runner_test.go
+++ b/js/runner_test.go
@@ -21,6 +21,7 @@
 package js
 
 import (
+	"bytes"
 	"context"
 	"crypto/tls"
 	"crypto/x509"
@@ -29,6 +30,7 @@ import (
 	stdlog "log"
 	"net"
 	"net/http"
+	"os"
 	"testing"
 	"time"
 
@@ -201,7 +203,8 @@ func TestOptionsPropagationToScript(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, expScriptOptions, r1.GetOptions())
 
-	r2, err := NewFromArchive(r1.MakeArchive(), lib.RuntimeOptions{Env: map[string]string{"expectedSetupTimeout": "3s"}})
+	r2, err := NewFromArchive(r1.MakeArchive(), lib.RuntimeOptions{Env: map[string]string{"expectedSetupTimeout": "1s"}})
+
 	require.NoError(t, err)
 	require.Equal(t, expScriptOptions, r2.GetOptions())
 
@@ -1342,6 +1345,86 @@ func TestInitContextForbidden(t *testing.T) {
 					t,
 					"GoError: "+test[2],
 					err.Error())
+			}
+		})
+	}
+}
+
+func TestArchiveRunningIntegraty(t *testing.T) {
+	tb := testutils.NewHTTPMultiBin(t)
+	defer tb.Cleanup()
+
+	fs := afero.NewMemMapFs()
+	require.NoError(t, afero.WriteFile(fs, "/home/somebody/test.json", []byte(`42`), os.ModePerm))
+	r1, err := New(&lib.SourceData{
+		Filename: "/script.js",
+		Data: []byte(tb.Replacer.Replace(`
+			let fput = open("/home/somebody/test.json");
+			export let options = { setupTimeout: "10s", teardownTimeout: "10s" };
+			export function setup() {
+				return JSON.parse(fput);
+			}
+			export default function(data) {
+				if (data != 42) {
+					throw new Error("incorrect answer " + data);
+				}
+			}
+		`)),
+	}, fs, lib.RuntimeOptions{})
+	require.NoError(t, err)
+
+	buf := bytes.NewBuffer(nil)
+	require.NoError(t, r1.MakeArchive().Write(buf))
+
+	arc, err := lib.ReadArchive(buf)
+	require.NoError(t, err)
+	r2, err := NewFromArchive(arc, lib.RuntimeOptions{})
+	require.NoError(t, err)
+
+	runners := map[string]*Runner{"Source": r1, "Archive": r2}
+	for name, r := range runners {
+		t.Run(name, func(t *testing.T) {
+			ch := make(chan stats.SampleContainer, 100)
+			err = r.Setup(context.Background(), ch)
+			require.NoError(t, err)
+			vu, err := r.NewVU(ch)
+			require.NoError(t, err)
+			err = vu.RunOnce(context.Background())
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestArchiveNotPanicing(t *testing.T) {
+	tb := testutils.NewHTTPMultiBin(t)
+	defer tb.Cleanup()
+
+	fs := afero.NewMemMapFs()
+	require.NoError(t, afero.WriteFile(fs, "/non/existant", []byte(`42`), os.ModePerm))
+	r1, err := New(&lib.SourceData{
+		Filename: "/script.js",
+		Data: []byte(tb.Replacer.Replace(`
+			let fput = open("/non/existant");
+			export default function(data) {
+			}
+		`)),
+	}, fs, lib.RuntimeOptions{})
+	require.NoError(t, err)
+
+	arc := r1.MakeArchive()
+	arc.Files = make(map[string][]byte)
+	r2, err := NewFromArchive(arc, lib.RuntimeOptions{})
+	require.NoError(t, err)
+
+	runners := map[string]*Runner{"Source": r1, "Archive": r2}
+	for name, r := range runners {
+		t.Run(name, func(t *testing.T) {
+			ch := make(chan stats.SampleContainer, 100)
+			_, err := r.NewVU(ch)
+			if name == "Source" {
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err)
 			}
 		})
 	}

--- a/lib/archive.go
+++ b/lib/archive.go
@@ -33,6 +33,8 @@ import (
 	"sort"
 	"strings"
 	"time"
+
+	"github.com/spf13/afero"
 )
 
 var volumeRE = regexp.MustCompile(`^([a-zA-Z]):(.*)`)
@@ -47,6 +49,18 @@ func NormalizeAndAnonymizePath(path string) string {
 	p = strings.Replace(p, "\\", "/", -1)
 	p = sharedRE.ReplaceAllString(p, `/nobody`)
 	return homeDirRE.ReplaceAllString(p, `$1/$2/nobody`)
+}
+
+type normalizedFS struct {
+	afero.Fs
+}
+
+func (m *normalizedFS) Open(name string) (afero.File, error) {
+	return m.Fs.Open(NormalizeAndAnonymizePath(name))
+}
+
+func (m *normalizedFS) OpenFile(name string, flag int, mode os.FileMode) (afero.File, error) {
+	return m.Fs.OpenFile(NormalizeAndAnonymizePath(name), flag, mode)
 }
 
 // An Archive is a rollup of all resources and options needed to reproduce a test identically elsewhere.
@@ -68,6 +82,8 @@ type Archive struct {
 	Scripts map[string][]byte `json:"-"` // included scripts
 	Files   map[string][]byte `json:"-"` // non-script resources
 
+	FS afero.Fs `json:"-"`
+
 	// Environment variables
 	Env map[string]string `json:"env"`
 }
@@ -78,6 +94,7 @@ func ReadArchive(in io.Reader) (*Archive, error) {
 	arc := &Archive{
 		Scripts: make(map[string][]byte),
 		Files:   make(map[string][]byte),
+		FS:      &normalizedFS{Fs: afero.NewMemMapFs()},
 	}
 
 	for {
@@ -133,6 +150,11 @@ func ReadArchive(in io.Reader) (*Archive, error) {
 		}
 
 		dst[name] = data
+
+		err = afero.WriteFile(arc.FS, name, data, os.ModePerm)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	return arc, nil

--- a/lib/archive_test.go
+++ b/lib/archive_test.go
@@ -78,6 +78,7 @@ func TestArchiveReadWrite(t *testing.T) {
 		assert.NoError(t, arc1.Write(buf))
 
 		arc2, err := ReadArchive(buf)
+		arc2.FS = nil
 		assert.NoError(t, err)
 		assert.Equal(t, arc1, arc2)
 	})
@@ -135,6 +136,7 @@ func TestArchiveReadWrite(t *testing.T) {
 			assert.NoError(t, arc1.Write(buf))
 
 			arc2, err := ReadArchive(buf)
+			arc2.FS = nil
 			assert.NoError(t, err)
 			assert.Equal(t, arc1Anon, arc2)
 		}

